docs/pages/database-access/guides/mysql-self-hosted.mdx

```diff
@@ -7,54 +7,20 @@ description: How to configure Teleport Database Access with self-hosted MySQL
 
```

Comment:
What do you think about adding:

```
(!docs/pages/includes/permission-warning.mdx!)

(!docs/pages/includes/backup-warning.mdx!)

## Create role and user

(!docs/pages/database-access/includes/db-user-roles.mdx!)
```

Reply:
We recommend creating at least one Teleport role specific for database access. 

And, users that wish to connect to a database through Teleport will want to have the ability to [impersonate](https://goteleport.com/docs/access-controls/guides/impersonation/) that role.

Below, we'll create both of these Role Templates to allow a user to connect to any database using any database account in one handy command:

```bash
cat > dbroles.yaml << 'EOF'
kind: role
version: v3
metadata:
  name: db
spec:
  allow:
    db_labels:
      '*': '*'
    db_names:
    - '*'
    db_users:
    - '*'
---
kind: role
version: v3
metadata:
  name: impdbrole
spec:
  allow:
    impersonate:
      users: ["Db"]
      roles: ["Db"]
EOF
```

```bash
tctl create -f dbroles.yaml
```

Create a user `db-user` with the defined database access like so:

```bash
tctl users add db-user --roles=access,impdbrole --logins=root,ubuntu,ec2-user
```

Reply:
The include ^ - I found that to be a bit confusing.

Reply:
I think this will break the flow of these guides. In general the way I currently designed them is:

1. First users configure their database instance (self-hosted or cloud).
2. Then they configure Teleport.

As such, creating Teleport users/roles in the beginning doesn't really make sense.

In addition there are a couple of inaccuracies in the version you proposed:

* Users don't have to impersonate `Db` to connect. It's only needed to generate secrets for self-hosted dbs in some cases (e.g. Cloud) - which is why it's at the beginning.
* These particular instructions are actually a bit obsolete - for new guides we don't ask them to create roles and use builtin `access` role we have. I need to update older guides as well but it's out of scope of this PR.

In general I'd like to move Teleport user creation to a partial like you suggested, though the difficulty is details differ a little bit between guides for different databases so I'm not sure yet best way to do it. I'll come back to it later, it's out of scope of this PR.

Reply:
The layout is currently a brief description about Teleport with several commands describe (although these are not explicit instructions or steps), then a partial walkthrough about configuring the database (with most steps omitted), then a bit about Teleport, then some more about the database(s), then some more about Teleport locally.

That's _verbatim_ what's on the page. Apart from that, in practice, the Guide interleaves multiple separate configuration steps and is difficult to complete.  Don't take my subjective opinion as the primary criteria for evaluation here:

_Bounce rates are 20% higher in this Section and it has the 4th highest exit rate of all content._

Reply:
@inertial-frame What do you propose?

Reply:
Don't mean for this to turn into a blocker (fan of the idea that we should merge changes if the PR's represent something non-breaking and strictly better even if there's something incomplete or could be improved - for example adding a section but not all of the planned articles, or updating some document even if everything's not 100% decided). 

I also want to avoid technical and editorial debt. We also have at least 2 weeks (and then some) before the 7.0 release (great job getting these changes in so early)!

---

For starters, I walked through each of these and documented the steps I needed to take to advance from beginning to end: https://docs.google.com/document/d/1C8CT-JSFpRcxPxfP1zX4a-LwTb5NKlWyG0mk4f6q5MI/edit# - I found these to be time-consuming and challenging - not necessarily because the articles were difficult but because configuring and installing a database is complex in its own right (plus all of the other items on top)!

The Teleport steps are actually the breeziest (again kudos on the great work). 

In walking through those steps, the one thing I found to be something we could improve on is the Teleport flow. We introduce a concept at the beginning but return to it later on. Also, no explicit example is given (a few small steps are).

That seemed to vastly simplify this process for me - it was not clear what roles were required, that the created user should be given the impersonation role, etc. The other items involved figuring out the best way to configure the DB's.

These opinions are not lines in the sand but I do think that step needs to be improved. Arguably, writing and reading a setting up DB's in a self-hosted capacity article (whether locally, on EC2, in a VM) should involve less domain knowledge (since you don't need to describe a lot of the IAM and configuration items - these are definite musts in production but can typically be skipped in an article whereas cloud configuration settings typically must be conveyed in an article - the latter is probably easier to use and get set up but the article format involves describing many additional pieces).

Reply:
Yeah I agree that instructions on how to install a self-hosted database should probably not be a part of these guides - each system is different and we wouldn't be able to cover them all. Also, installing a database server is not actually the part folks struggle with the most (usually, IAM stuff is the most tricky).

One issue about the current self-hosted flow that I've just realized is that we ask the users to use `tctl auth sign` to produce certificates - which requires a Teleport cluster, but the instructions of installing Teleport actually follow much later... We could tweak it a bit like this:

1. Install and configure Teleport Auth/Proxy - OR - Instructions for Teleport Cloud users.
2. Configure Postgres/MySQL database, which involves:
  2.1 Using `tctl auth sign` to produce secrets.
  2.2 Use those secrets to configure mTLS on the database.
  2.3 Create database user, etc.
3. Then proceed to configuring Teleport Database service.
4. Finally, connect.

So basically installing/configuring Teleport cluster moves to first step and the rest follows - one thing kinds of leads to another. What do you think? I could try and apply this change to the new Mongo guide here (since it's basically self-hosted) and do other guides later as well.

Reply:
@inertial-frame I have refactored MongoDB guide. I think it turned out quite nicer now, I rearranged the flow as described above, simplified some things and moved some things into partials. Here's a brief summary:

* Explicitly mentioned steps and prerequisites at the beginning like other guides do.
* Rearranged the flow to start with Teleport - it flows much better now (IMO).
* Made "create user" section more generic and moved into a partial - so we can reuse in other guides too.
* Removed "start with configuration file" section entirely - I think it was only complicating things.
* Added more clarify around tctl auth usage for self-hosted Teleport vs Cloud.

Here's the commit with the changes: https://github.com/gravitational/teleport/pull/7350/commits/d3b8ebc903553f186c8c9c7e2dfc3833845aa4cd. Let me know what you think.